bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java

```diff
@@ -278,62 +364,23 @@ private synchronized void updateModuleHandlerFactoryMap(Collection<String> remov
      * @param rule a rule which has to be added.
      * @param isEnabled specifies the rule to be added as disabled or not.
      */
-    protected void addRule(Rule rule, boolean isEnabled) {
+    protected void addRule(Rule rule, RuleStatusInfo statusInfo) {
```

Comment:
Why this change? It feels wrong to pass a statusInfo with a rule - the status should always only be determined by the rule engine.

Reply:
The RuleRegistry interface has several methods connected to the status: getStatusInfo, getStatus, isEnabled, setEnabled. With this PR we are decoupling the RuleRegistry and the RuleEngine, so now the registry should be able to work without the engine. The RuleRegistry is responsible for disabling/enabling the rules and for the resolving & normalization of the configurations - which changes the status of the rule. When the RuleEngine is missing, the RuleRegistry determines the status based on these functions - the status can be DISABLED, INITIALIZING or UNINITIALIZED. When the RuleEngine appears, the RuleRegistry must pass this initial status - so the isEnabled flag is not enough. The engine can then further change it.

Reply:
This sounds as if you try to workaround a startup order issue by exposing internals on the API... If you look at the (old) API, I hope you agree that this is what the user expects it to look like - so we should keep it that way.

Regarding the lifecyle issues, see [my comment in July](https://github.com/eclipse/smarthome/issues/3635#issuecomment-318335109). If the RuleEngine has not registered as a listener on the registry, so be it. The Registry should not be responsible for enabling/disabling, etc. as it is not the rule engine. The Registry should merely serve as an API facade and pass on those requests. If there is no one to pass it on to, so be it.

Reply:
But how the Rule Registry will receive a resolved rule after all and if it is responsible for resolving, how the RuleEngine will now the status of the rule after resolving?

Reply:
> I would expect the rule engine to treat the passed Rule object as immutable and not to do any changes to it.

these are your words and I'm taking that into consideration